src/gt4py/next/iterator/transforms/global_tmps.py

```diff
@@ -127,56 +177,122 @@ def split_closures(node: ir.FencilDefinition) -> FencilWithTemporaries:
     3. Extract lifted function class as new closures with the previously created temporary as output.
     The closures are processed in reverse order to properly respect the dependencies.
     """
-    tmps: list[ir.SymRef] = []
-
-    def handle_arg(arg):
-        """Handle arguments of closure calls: extract lifted function calls.
-
-        Lifted function calls, do:
-        1. Replace the call by a new symbol ref, put this into `tmps`.
-        2. Put the ‘unlifted’ function call to the stack of stencil calls that still have to be
-        processed.
-        """
-        if isinstance(arg, ir.SymRef):
-            return arg
-        if (
-            isinstance(arg, ir.FunCall)
-            and isinstance(arg.fun, ir.FunCall)
-            and arg.fun.fun == ir.SymRef(id="lift")
-        ):
-            assert len(arg.fun.args) == 1
-            ref = ir.SymRef(id=f"_gtmp_{len(tmps)}")
-            tmps.append(ir.Sym(id=ref.id))
-            unlifted = ir.FunCall(fun=arg.fun.args[0], args=arg.args)
-            stencil_stack.append((ref, unlifted))
-            return ref
-        raise AssertionError()
+    uid_gen_tmps = UIDGenerator(prefix="_tmp")
+
+    type_inference.infer_all(node, offset_provider=offset_provider, save_to_annex=True)
 
-    closures = []
+    tmps: list[ir.Sym] = []
+
+    closures: list[ir.StencilClosure] = []
     for closure in reversed(node.closures):
-        wrapped_stencil = ir.FunCall(fun=closure.stencil, args=closure.inputs)
-        popped_stencil = PopupTmps().visit(wrapped_stencil)
+        closure_stack: list[ir.StencilClosure] = [closure]
+        while closure_stack:
+            current_closure: ir.StencilClosure = closure_stack.pop()
+
+            if current_closure.stencil == im.ref("deref"):
+                closures.append(current_closure)
+                continue
+
+            # TODO(tehrengruber): extracting from a scan is untested.
```

Comment:
Maybe should we disable the scan case for now until we have a test?

Reply:
I've added a test.